### PR TITLE
A WIP commit on the new refactor for support of Python and other features

### PR DESCRIPTION
**Pull request by garnaat**
Not ready to be merged but anyone who wants to review it, I would be much obliged.

### CODE REVIEW

File: kappa/context.py [hunk 1]
```diff
@@ -23,39 +25,60 @@
...
         self.event_sources = []
         self._create_event_sources()
 
+    def _load_cache(self):
+        self.cache = {}
+        if os.path.isdir('.kappa'):
+            cache_file = os.path.join('.kappa', 'cache')
+            if os.path.isfile(cache_file):
+                with open(cache_file, 'rb') as fp:
+                    self.cache = yaml.load(fp)
```
Comment (kapilt) [inline kappa/context.py]: might want to do a utility function here, to use the c impl, and to use a SafeLoader, as default yaml encoder does arbitrary object serialization/de-serialization (aka pickle), which turns cache files into attack vectors. also is there a time limit on the cache?

File: kappa/event_source.py [hunk 2]
```diff
@@ -73,12 +75,37 @@ def add(self, function):
...
+            response = self._lambda.call(
+                'update_event_source_mapping',
+                FunctionName=function.name,
+                Enabled=self.enabled
+            )
+            LOG.debug(response)
+        except Exception:
+            LOG.exception('Unable to enable event source')
+
+    def disable(self, function):
```
Comment (kapilt) [inline kappa/event_source.py]: i'd consider just a set_enabled(self, function, enabled=True) to reduce the duplication.

**Comment by kapilt** [comment]
Looks in good in general, the recording and mock replay for testing is interesting, but it doesn't seem to be wired up per se, mock aws is using module level attributes for response value. for general purpose usage on aws client, an option to not auto paginate seems appropriate. added a few inline comments.

**Comment by garnaat** [comment]
Thanks for the comments.

Regarding the recording/mock playback I have a similar mechanism in skew and I'm considering breaking that out into a separate library to avoid duplication.  Or perhaps something like vcr or responses could do the same thing.  I'm not interested in re-inventing the wheel but I seem to face this testing issue on every library I write and I'd like to find a reasonable solution.

**Comment by kapilt** [comment]
A separate recorder/proxy/mock playback lib sounds good, i'd be interested in helping out. Its a common issue. I've been making do with ftests, mock, and unit tests atm, but its affecting structure of the code to keep the mocking sane for a given test. stub style with moto had appeal, but its too fragile and incompatible while straddling the boto 2/3 divide.

**Comment by garnaat** [comment]
I'll try to get a repo up with a start of that today.  I'm kind of swamped at the moment but I would like to get this kappa version out there asap.

**Comment by hunt3r** [comment]
I am very much interested in this branch.  We're starting a new Lambda project and I was about to write a bunch of tasks to manage the workflow.  These seems to fit the bill once the Python support is added.

**Comment by garnaat** [comment]
I encourage you to check it out now.  I have been using it for a few weeks and its working for me but I know my use case is narrow and specific.  I would love to get hands-on feedback and it would provide you with an opportunity to get kappa to work exactly the way you want it to work.

**Comment by kapilt** [comment]
Feels like this branch should have been merged a while back, its getting new unrelated development going on, and its unclear now where to contribute new features too. also per travis errors, its missing restapi module..

**Comment by garnaat** [comment]
I agree.  I don't think I have ever had a PR open this long before.

The bottom line is that I am using kappa everyday and I'm able to make changes that I need in this branch but I just haven't had the time yet to do all of the work (e.g. tests, docs, etc.) required to merge this into develop.

I hope to be able to do that soon.

**Comment by kapilt** [comment]
Yeah, the question is where to send additional features/pull requests on contributions? Target this branch or mainline? Frankly i'll probably route through our corporate contrib process which means a few weeks delay in any case, so not urgent.

**Comment by garnaat** [comment]
This is the branch to target for changes.  It will get merged soon and will be the new version of kappa.

**Pull request merged**